### ISSUE TITLE
WinForms - .NET 7.0 Designer unable to load

### ISSUE THREAD

**Issue by davidei1955** [cefsharp/CefSharp#4424]
- What version of the product are you using?  **CefSharp.Winforms.NETCore Vers. 110.0.0300**

- What version are you using?  **Nuget CefSharp.Winforms.NETCore Vers. 110.0.0300**

- What architecture x86 or x64?  **I tried all 3:  AnyCPU, x64, x86, **
	
- What version of .Net?  **.NET 7.0**
    
- On what operating system?   **Windows 11 Pro x64**

- Are you using `WinForms`, `WPF` or `OffScreen`?  **WinForms**

**What steps will reproduce the problem?**

1. Create new Windows Forms App: Framework: .NET 7.0 in Visual Studio 2022

2. Open WinForms Designer and verify that it works:  **rt-click Form1.cs > View Designer**

3. NuGet Add: CefSharp.Winforms.NETCore to project: Latest Stable 110.0.0300

4. Try to open WinFoms Designer and notice that it no longer opens

5. View the error in the Output Window:  - Show output from: Windows Forms:
`   [18:27:20.806087] fail: [WinFormsTest]: Failed to load assembly: 
\\?\C:\Users\DavidEi\AppData\Local\Microsoft\VisualStudio\17.0_242dc2dc\WinFormsDesigner\joxhj40u.o04\UserAppData\Ijwhost.dll
System.BadImageFormatException: Bad IL format. The format of the file is invalid.
...`

6. In Configuration Manager, change the Platform to x64 and notice that the Designer still won't open.

7. In Configuration Manager, change the Platform to x86 and notice that the Designer still won't open.

8. Open NuGet and remove : CefSharp.Winforms.NETCore from the project

9. Close Visual Studio, reopen it, and rebuild the project.

10. Notice that the designer opens correctly, irrespective of the Platform setting

- Does the problem reproduce using the [MinimalExample](https://github.com/cefsharp/CefSharp.MinimalExample)?
   **CefSharp.MinimalExample exhibits a similar problem when trying to open the designer on _BrowserForm.cs_
      I was able to get the designer to open when the Platform setting was set to x64 (required a VS restart).
      MinimalExample initially opens with AnyCPU**

- What is the expected output? What do you see instead? 
  **Expected the Forms Designer to open. Displays 'Loading designer...' forever, instead.
     Output Window (Show output from _Windows Forms_) displays an exception**

- Does the cef log provide any relevant information? (By default there should be a debug.log file in your bin directory) 
   **There is no [xxx].log in my project file tree**

- Any other background information that's relevant? 
  **This is related to, but not the same as, _Bug 4019_ (WinForms - Add support for OOP Designer (VS2022). I'm not trying to use the CefSharp browser as a Design-Time control, I just want to open the Designer to  view Winforms forms. I can't temporarily remove CefSharp.Winforms.NETCore from the project, it won't compile without it.
This problem does not exist in .Net Framework 4.8. I encountered it when porting an old .Net 4.8 app to .NET 7.0.**

**Comment by davidei1955** [comment]
Adding: `

    <PropertyGroup Condition="'$(PlatformTarget)' == 'x86'">
        <RuntimeIdentifier Condition="'$(RuntimeIdentifier)' == ''">win-x86</RuntimeIdentifier>
        <SelfContained Condition="'$(SelfContained)' == ''">false</SelfContained>
    </PropertyGroup>

    <PropertyGroup Condition="'$(PlatformTarget)' == 'x64'">
        <RuntimeIdentifier Condition="'$(RuntimeIdentifier)' == ''">win-x64</RuntimeIdentifier>
        <SelfContained Condition="'$(SelfContained)' == ''">false</SelfContained>
    </PropertyGroup>
`
And changing 'Platforms' to:

`<Platforms>x64;x86</Platforms>`

In my .csproj file resolved the problem! 

I suspect that changing MinimalExample so that it comes up with its Platform set to either x86 or x64, rather than AnyCPU, would allow you to open BrowserForm.cs in the designer.

Also, I had to hand-edit the .csproj to modify the `<Platforms>` tag. The Configuration Manager didn't seem to do it reliably.

Have you seen: [https://devblogs.microsoft.com/dotnet/custom-controls-for-winforms-out-of-process-designer/](url)? It's pretty recent and discusses how to create custom control designers for the new Winforms OOP designer. 

You can close this issue. **Thanks for your help!**

**Comment by amaitland** [comment]
Firstly thank you for your support, greatly appreciated 👍 

Do you happen to know what version of `VS2022` you were testing with?

A quick test with `17.5.1` and the designer was loading, there were exceptions regarding `CefSharp.Core.Runtime.dll`, it did at least appear to load.

My general experience with the `OOP Designer` is that it appears to be a little flakey, I've seen it fail to open the designer for a brand new `WinForms .Net 7` project.

> Have you seen: https://devblogs.microsoft.com/dotnet/custom-controls-for-winforms-out-of-process-designer/ It's pretty recent and discusses how to create custom control designers for the new WinForms OOP designer.

Thanks for the link, this appears to be slightly newer than the content previously linked in https://github.com/dotnet/winforms/discussions/7073

The [WinForms Designer SDK](https://www.nuget.org/packages/Microsoft.WinForms.Designer.SDK/1.1.0-prerelease-preview3.22076.5) itself is still in preview and hasn't been updated in quite some time.

**Comment by davidei1955** [comment]
VS version info:

Microsoft Visual Studio Community 2022

Version 17.4.3

VisualStudio.17.Release/17.4.3+33205.214

Microsoft .NET Framework

Version 4.8.09032

 

Installed Version: Community

 

Visual C++ 2022   00482-90000-00000-AA259

Microsoft Visual C++ 2022

 

ASP.NET and Web Tools   17.4.326.54890

ASP.NET and Web Tools

 

Azure App Service Tools v3.0.0   17.4.326.54890

Azure App Service Tools v3.0.0

 

C# Tools   4.4.0-6.22580.4+d7a61210a88b584ca0827585ec6e871c6b1c5a14

C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

 

Microsoft JVM Debugger   1.0

Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

 

NuGet Package Manager   6.4.0

NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

 

Syntax Visualizer   1.0

An extension for visualizing Roslyn SyntaxTrees.

 

Test Adapter for Boost.Test   1.0

Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

 

Test Adapter for Google Test   1.0

Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

 

TypeScript Tools   17.0.10921.2001

TypeScript Tools for Microsoft Visual Studio

 

Visual Basic Tools   4.4.0-6.22580.4+d7a61210a88b584ca0827585ec6e871c6b1c5a14

Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

 

Visual F# Tools   17.4.0-beta.22512.4+525d5109e389341bb90b144c24e2ad1ceec91e7b

Microsoft Visual F# Tools

 

Visual Studio IntelliCode   2.2

AI-assisted development for Visual Studio.

 

I'm pretty sure the designer startup problems were related to the designer not loading the correct arch for the CefSharp dll. 

 

I recently ported an old, moderately large, Winforms app from .Net 4.8 to .NET 7.0. It has about 15 Forms. I was able to open and edit them all, once I got the arch straightened out and added a RuntimeIdentifier. I had to manually edit the .csproj config files to target the x64 platform, however. Configuration Manager wouldn't modify the <Platforms> tag. Now, the x86 & x64 configurations both work with the designer. I have no confidence in AnyCPU (the default, unfortunately). The app uses Cef and ImageMagick C++ dlls (with CefSharp and Magick.Net wrappers).

This may be user error since I'm not very familiar with the syntax, but I had to remove several of the old projects, create new projects, and copy the files from the old projects into the new projects, to get many of the projects (especially librraries) to build. It kept complaining about incompatible target frameworks, despite any chages I'd make to the <TargetFrameworks> tags. MinimalExample wouldn't even build for me till I recreated the project. I only cared that it targeted .NET 7.0, so I chose that when creating the new replacement projects. This threw away a lot of unnecessary .Net 4.x cruft and may have contributed to fixing the designer startup problems (as well as adding a RuntimeIdentifier).

 

FWIW, Once I got the OOP Designer to open, I found it pretty stable with VS 17.4 & .NET 7.0. It only crashed on me a couple of times over the course of 2 weeks. It was unusable in .NET 5.0, and was usable but crashed a lot in .NET 6.0 (I'm not sure what version of VS I was using - whatever was current in the .NET 5/6 timeframes).

 

Feel free to add this email as a comment to my bug report issue, if you think its appropriate.

 

I see there's a new release of VS, I'll update and do a bit of testing with your new browser and let you know how it goes.

 

Best regards,

 

David Ei

   Email:     ***@***.***> ***@***.***

   Home:  (734) 449-2404

   Cell:       (734) 353-9542
